client/x/evmstaking/keeper/msg_queue.go

```diff
@@ -0,0 +1,79 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	slashingtypes "github.com/cosmos/cosmos-sdk/x/slashing/types"
+	stype "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/piplabs/story/client/x/evmstaking/types"
+	"github.com/piplabs/story/lib/errors"
+	"github.com/piplabs/story/lib/log"
+)
+
+// EnqueueMsg enqueues a message to the queue of the current epoch.
+func (k Keeper) EnqueueMsg(ctx context.Context, msg types.QueuedMessage) error {
+	return k.MessageQueue.Enqueue(ctx, msg)
+}
+
+// ProcessAllMsgs returns the set of messages queued in a given epoch.
+func (k Keeper) ProcessAllMsgs(ctx context.Context) error {
+	iterator, err := k.MessageQueue.Iterate(ctx)
+	if err != nil {
+		return errors.Wrap(err, "message queue iterator")
+	}
+
+	for ; iterator.Valid(); iterator.Next() {
+		qMsg, err := iterator.Value()
+		if err != nil {
+			return errors.Wrap(err, "get value of message queue")
+		}
+
+		if err := k.processMsg(ctx, &qMsg); err != nil {
+			log.Warn(ctx, "Failed to process queued message", err, "tx_id", string(qMsg.TxId))
+			return errors.Wrap(err, "process queued message")
+		}
+	}
+
+	return nil
+}
+
+// processMsg processes queues message depending on the type of message.
+func (k Keeper) processMsg(ctx context.Context, msg *types.QueuedMessage) error {
+	var (
+		unwrappedMsgWithType sdk.Msg
+		err                  error
+	)
+	unwrappedMsgWithType, err = msg.UnwrapToSdkMsg()
+	if err != nil {
+		return errors.Wrap(err, "unwrap queued msg to sdk msg")
+	}
+
+	switch unwrappedMsg := unwrappedMsgWithType.(type) {
+	case *stype.MsgCreateValidator:
+		if err := k.ProcessCreateValidatorMsg(ctx, unwrappedMsg); err != nil {
+			return errors.Wrap(err, "process CreateValidator msg")
+		}
+	case *stype.MsgDelegate:
+		if err := k.ProcessDepositMsg(ctx, unwrappedMsg); err != nil {
```

Comment:
In the future, maybe we can add an improvement where deposits are handled per block but only go into effect at the end of each epoch.